File: src/Titan.D3D11/Device/D3D11Device.cs
using System;
using Titan.D3D11.Bindings;
using Titan.D3D11.Bindings.Models;
using Titan.D3D11.InfoQueue;
using Titan.Windows;

namespace Titan.D3D11.Device
{
    internal class D3D11Device : ID3D11Device
    {
        private readonly IntPtr _handle;
        public D3DFeatureLevel FeatureLevel { get; }
        public ID3D11SwapChain SwapChain { get; }
        public ID3D11DeviceContext Context { get; }

        public D3D11Device(IntPtr handle, D3DFeatureLevel featureLevel, ID3D11SwapChain swapChain, ID3D11DeviceContext context)
        {
            _handle = handle;
            FeatureLevel = featureLevel;
            SwapChain = swapChain;
            Context = context;
        }

        public ID3D11RenderTargetView CreateRenderTargetView(ID3D11Resource resource)
        {
            HRESULT result;
            IntPtr renderTargetView;
            unsafe
            {
                result = D3D11DeviceBindings.CreateRenderTargetView_(_handle, resource.Handle, (D3D11RenderTargetViewDesc*)null, out renderTargetView);
            }
            result.Check(nameof(D3D11Device));
            return new D3D11RenderTargetView(renderTargetView);
        }

        public ID3D11InfoQueue CreateInfoQueue()
        {
            var result = D3D11CommonBindings.QueryInterface_(_handle, D3D11Resources.D3D11InfoQueue, out var infoQueue);
            result.Check(nameof(D3D11Device));
            return new D3D11InfoQueue(infoQueue);
        }

        public ID3D11Buffer CreateBuffer(in D3D11BufferDesc desc, D3D11SubresourceData? subresourceData = null)
        {
            HRESULT result;
            IntPtr buffer;
            unsafe
            {
                fixed (D3D11BufferDesc* bufferDesc = &desc)
                {
                    if (subresourceData.HasValue)
                    {
                        var subResource = subresourceData.Value;
                        result = D3D11DeviceBindings.CreateBuffer_(_handle, bufferDesc, &subResource, out buffer);
                    }
                    else
                    {
                        result = D3D11DeviceBindings.CreateBuffer_(_handle, bufferDesc, null, out buffer);
                    }
                }
            }
            result.Check(nameof(D3D11Device));
            return new D3D11Buffer(buffer);
        }

        public ID3D11VertexShader CreateVertexShader(ID3DBlob blob)
        {
            return CreateVertexShader(blob.GetBufferPointer(), blob.GetBufferSize());
        }

        public ID3D11VertexShader CreateVertexShader(IntPtr buffer, UIntPtr size)
        {
            var result = D3D11DeviceBindings.CreateVertexShader_(_handle, buffer, size , IntPtr.Zero, out var vertexShader);
            result.Check(nameof(D3D11Device));
            return new D3D11VertexShader(vertexShader);
        }

        public ID3D11PixelShader CreatePixelShader(ID3DBlob blob)
        {
            return CreatePixelShader(blob.GetBufferPointer(), blob.GetBufferSize());
        }

        public ID3D11PixelShader CreatePixelShader(IntPtr buffer, UIntPtr size)
        {
            var result = D3D11DeviceBindings.CreatePixelShader_(_handle, buffer, size, IntPtr.Zero, out var pixelShader);
            result.Check(nameof(D3D11Device));
            return new D3D11PixelShader(pixelShader);
        }

        public ID3D11InputLayout CreateInputLayout(in D3D11InputElementDesc[] descs, ID3DBlob blob)
        {
            return CreateInputLayout(descs, blob.GetBufferPointer(), blob.GetBufferSize());
        }

        public ID3D11InputLayout CreateInputLayout(in D3D11InputElementDesc[] descs, IntPtr buffer, UIntPtr size)
        {
            var result = D3D11DeviceBindings.CreateInputLayout_(_handle, descs, (uint)descs.Length, buffer, size, out var inputLayout);
            result.Check(nameof(D3D11Device));
            return new D3D11InputLayout(inputLayout);
        }
        public ID3D11DepthStencilState CreateDepthStencilState(in D3D11DepthStencilDesc depthDesc)
        {
            HRESULT result;
            IntPtr depthStencilState;
            unsafe
            {
                fixed (D3D11DepthStencilDesc* pointer = &depthDesc)
                {
                    result = D3D11DeviceBindings.CreateDepthStencilState_(_handle, pointer, out depthStencilState);
                }
            }
            result.Check(nameof(D3D11Device));
            return new D3D11DepthStencilState(depthStencilState);
        }

        public ID3D11Texture2D CreateTexture2D(in D3D11Texture2DDesc texture2DDesc, D3D11SubresourceData? initialData = null)
        {
            HRESULT result;
            IntPtr texture2D;
            unsafe
            {
                fixed (D3D11Texture2DDesc* texture2DPointer = &texture2DDesc)
                {
                    if (initialData.HasValue)
                    {
                        var subResourceData = initialData.Value; // copies the value, maybe look at other ways to do this.
                        result = D3D11DeviceBindings.CreateTexture2D_(_handle, texture2DPointer, &subResourceData, out texture2D);
                    }
                    else
                    {
                        result = D3D11DeviceBindings.CreateTexture2D_(_handle, texture2DPointer, null, out texture2D);
                    }
                }
            }
            result.Check(nameof(D3D11Device));
            return new D3D11Texture2D(texture2D);
        }

        public ID3D11DepthStencilView CreateDepthStencilView(ID3D11Resource resource, in D3D11DepthStencilViewDesc desc)
        {
            HRESULT result;
            IntPtr stencilView;
            unsafe
            {
                fixed (D3D11DepthStencilViewDesc* viewDescPointer = &desc)
                {
                    result = D3D11DeviceBindings.CreateDepthStencilView_(_handle, resource.Handle, viewDescPointer, out stencilView);
                }
            }
            result.Check(nameof(D3D11Device));
            return new D3D11DepthStencilView(stencilView);
        }

        public ID3D11SamplerState CreateSamplerState(in D3D11SamplerDesc desc)
        {
            var result = D3D11DeviceBindings.CreateSamplerState_(_handle, desc, out var sampleState);
            result.Check(nameof(D3D11Device));
            return new D3D11SamplerState(sampleState);
        }

        public ID3D11ShaderResourceView CreateShaderResourceView(ID3D11Resource resource, D3D11ShaderResourceViewDesc desc)
        {
            var result = D3D11DeviceBindings.CreateShaderResourceView_(_handle, resource.Handle, desc, out var resourceView);
            result.Check(nameof(D3D11Device));
            return new D3D11ShaderResourceView(resourceView);
        }

        public ID3D11BlendState CreateBlendState(in D3D11BlendDesc desc)
        {
            var result = D3D11DeviceBindings.CreateBlendState_(_handle, desc, out var blendState);
            result.Check(nameof(D3D11Device));
            return new D3D11BlendState(blendState);
        }

        public ID3D11DeviceContext CreateDeferredContext(uint contextFlags = 0)
        {
            var result = D3D11DeviceBindings.CreateDeferredContext_(_handle, contextFlags, out var deferredContext);
            result.Check(nameof(D3D11Device));
            return new D3D11DeviceContext(deferredContext);
        }

        public D3DFeatureLevel GetFeatureLevel()
        {
            return D3D11DeviceBindings.GetFeatureLevel_(_handle);
        }

        public void Dispose()
        {
            SwapChain.Dispose();
            Context.Dispose();
            D3D11CommonBindings.ReleaseComObject(_handle);
        }
    }
}

File: src/Titan.D3D11.Bindings/D3D11Device.cpp
#include "common.h"

EXPORT HRESULT D3D11CreateDevice_(
	IDXGIAdapter*			pAdapter,
	D3D_DRIVER_TYPE         driverType,
	HMODULE                 software,
	UINT                    flags,
	const D3D_FEATURE_LEVEL* pFeatureLevels,
	UINT                    featureLevels,
	UINT                    SDKVersion,
	ID3D11Device**			ppDevice,
	D3D_FEATURE_LEVEL*		pFeatureLevel,
	ID3D11DeviceContext**	ppImmediateContext
)
{
	return D3D11CreateDevice(pAdapter, driverType, software, flags, pFeatureLevels, featureLevels, SDKVersion, ppDevice, pFeatureLevel, ppImmediateContext);
}

EXPORT HRESULT D3D11CreateDeviceAndSwapChain_(
	IDXGIAdapter*				pAdapter,
	D3D_DRIVER_TYPE				driverType,
	HMODULE						software,
	UINT						flags,
	const D3D_FEATURE_LEVEL*	pFeatureLevels,
	UINT						featureLevels,
	UINT						SDKVersion,
	DXGI_SWAP_CHAIN_DESC*		pSwapChainDesc,
	IDXGISwapChain**			ppSwapChain,
	ID3D11Device**				ppDevice,
	D3D_FEATURE_LEVEL*			pFeatureLevel,
	ID3D11DeviceContext**		ppImmediateContext
)
{
	return D3D11CreateDeviceAndSwapChain(pAdapter, driverType, software, flags, pFeatureLevels, featureLevels, SDKVersion, pSwapChainDesc, ppSwapChain, ppDevice, pFeatureLevel, ppImmediateContext);
}


EXPORT HRESULT CreateRenderTargetView_(
	ID3D11Device*				device,
	ID3D11Resource*				pResource,
	const D3D11_RENDER_TARGET_VIEW_DESC* pDesc,
	ID3D11RenderTargetView**	ppRTView
)
{
	return device->CreateRenderTargetView(pResource, pDesc, ppRTView);
}
EXPORT HRESULT CreateBuffer_(
	ID3D11Device* device,
	const D3D11_BUFFER_DESC* pDesc,
	const D3D11_SUBRESOURCE_DATA* pInitialData,
	ID3D11Buffer** ppBuffer)
{
	return device->CreateBuffer(pDesc, pInitialData, ppBuffer);
}

EXPORT HRESULT CreateVertexShader_(
	ID3D11Device * device,
	const void* pShaderBytecode,
	SIZE_T BytecodeLength,
	ID3D11ClassLinkage* pClassLinkage,
	ID3D11VertexShader** ppVertexShader
)
{
	return device->CreateVertexShader(pShaderBytecode, BytecodeLength, pClassLinkage, ppVertexShader);
}


EXPORT HRESULT CreatePixelShader_(
	ID3D11Device* device,
	const void* pShaderBytecode,
	SIZE_T BytecodeLength,
	ID3D11ClassLinkage* pClassLinkage,
	ID3D11PixelShader** ppPixelShader
)
{
	return device->CreatePixelShader(pShaderBytecode, BytecodeLength, pClassLinkage, ppPixelShader);
}

EXPORT HRESULT CreateInputLayout_(
	ID3D11Device* device,
	const D3D11_INPUT_ELEMENT_DESC* pInputElementDescs,
	UINT NumElements,
	const void* pShaderBytecodeWithInputSignature,
	SIZE_T bytecodeLength,
	ID3D11InputLayout** ppInputLayout
)
{
	return device->CreateInputLayout(pInputElementDescs, NumElements, pShaderBytecodeWithInputSignature, bytecodeLength, ppInputLayout);
}

EXPORT HRESULT CreateDepthStencilState_(
	ID3D11Device* device,
	const D3D11_DEPTH_STENCIL_DESC* pDepthStencilDesc,
	ID3D11DepthStencilState** ppDepthStencilState
)
{
	return device->CreateDepthStencilState(pDepthStencilDesc, ppDepthStencilState);
}

EXPORT HRESULT CreateTexture2D_(
	ID3D11Device* device,
	const D3D11_TEXTURE2D_DESC* pDesc,
	const D3D11_SUBRESOURCE_DATA* pInitialData,
	ID3D11Texture2D** ppTexture2D
) 
{
	return device->CreateTexture2D(pDesc, pInitialData, ppTexture2D);
}


EXPORT HRESULT CreateDepthStencilView_(
	ID3D11Device* device,
	ID3D11Resource* pResource,
	const D3D11_DEPTH_STENCIL_VIEW_DESC* pDesc,
	ID3D11DepthStencilView** ppDepthStencilView
)
{
	return device->CreateDepthStencilView(pResource, pDesc, ppDepthStencilView);
}
